staging/src/k8s.io/apiserver/pkg/endpoints/audit_test.go

```diff
@@ -0,0 +1,326 @@
+			},
+			selfLinker,
+			200,
+			1,
+			[]eventCheck{
+				requestBodyMatches(0, "DeleteOptions"),
+				responseBodyMatches(0, ""),
+			},
+		},
+		{
+			"update",
+			func(server string) (*http.Request, error) {
+				return http.NewRequest("PUT", server+"/"+prefix+"/"+testGroupVersion.Group+"/"+testGroupVersion.Version+"/namespaces/other/simple/c", bytes.NewBuffer(simpleCPrimeJSON))
+			},
+			selfLinker,
+			200,
+			1,
+			[]eventCheck{
+				requestBodyIs(0, string(simpleCPrimeJSON)),
+				responseBodyMatches(0, `{.*"bla".*}`),
+			},
+		},
+		{
+			"update-wrong-namespace",
+			func(server string) (*http.Request, error) {
+				return http.NewRequest("PUT", server+"/"+prefix+"/"+testGroupVersion.Group+"/"+testGroupVersion.Version+"/namespaces/default/simple/c", bytes.NewBuffer(simpleCPrimeJSON))
+			},
+			selfLinker,
+			400,
+			1,
+			[]eventCheck{
+				requestBodyIs(0, string(simpleCPrimeJSON)),
+				responseBodyIs(0, ""),
+			},
+		},
+		{
+			"patch",
+			func(server string) (*http.Request, error) {
+				req, _ := http.NewRequest("PATCH", server+"/"+prefix+"/"+testGroupVersion.Group+"/"+testGroupVersion.Version+"/namespaces/other/simple/c", bytes.NewReader([]byte(`{"labels":{"foo":"bar"}}`)))
+				req.Header.Set("Content-Type", "application/merge-patch+json; charset=UTF-8")
+				return req, nil
+			},
+			&setTestSelfLinker{
+				t:           t,
+				expectedSet: "/" + prefix + "/" + testGroupVersion.Group + "/" + testGroupVersion.Version + "/namespaces/other/simple/c",
+				name:        "c",
+				namespace:   "other",
+			},
+			200,
+			1,
+			[]eventCheck{
+				requestBodyIs(0, `{"labels":{"foo":"bar"}}`),
+				responseBodyMatches(0, `"name":"c".*"labels":{"foo":"bar"}`),
+			},
+		},
+		{
+			"watch",
+			func(server string) (*http.Request, error) {
+				return http.NewRequest("GET", server+"/"+prefix+"/"+testGroupVersion.Group+"/"+testGroupVersion.Version+"/namespaces/other/simple?watch=true", nil)
+			},
+			&setTestSelfLinker{
+				t:           t,
+				expectedSet: "/" + prefix + "/" + testGroupVersion.Group + "/" + testGroupVersion.Version + "/namespaces/other/simple",
+				namespace:   "other",
+			},
+			200,
+			2,
+			[]eventCheck{
+				requestBodyMatches(0, ""),
+				responseBodyMatches(0, ""),
+			},
+		},
+	} {
+		sink := &fakeAuditSink{}
+		handler := handleInternal(map[string]rest.Storage{
+			"simple": &SimpleRESTStorage{
+				list: []genericapitesting.Simple{
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "a", Namespace: "other"},
+						Other:      "foo",
+					},
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "b", Namespace: "other"},
+						Other:      "foo",
+					},
+				},
+				item: genericapitesting.Simple{
+					ObjectMeta: metav1.ObjectMeta{Name: "c", Namespace: "other", UID: "uid"},
+					Other:      "foo",
+				},
+			},
+		}, admissionControl, selfLinker, sink)
+
+		server := httptest.NewServer(handler)
```

Comment:
It wouldn't harm to have real integration tests in addition, this is really something between unit and integration.